From e6c8094e648099630f9ea7996efa0fd1860385bc Mon Sep 17 00:00:00 2001
From: Peter Chen <peter.chen@nxp.com>
Date: Fri, 13 Apr 2018 16:12:52 +0800
Subject: [PATCH 3702/5242] MLK-17631-5 usb: cdns3: gadget: set selfpower flag
 for usb_gadget

commit  2975d76929cc28d13bc9b3ed991ed9daa59d8153 from
https://source.codeaurora.org/external/imx/linux-imx.git

Implement selfpower setting between USB gadget core and
controller driver

Reviewed-by: Jun Li <jun.li@nxp.com>
Signed-off-by: Peter Chen <peter.chen@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/cdns3/gadget.c |    7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/drivers/usb/cdns3/gadget.c b/drivers/usb/cdns3/gadget.c
index 180bee1..52375a9 100644
--- a/drivers/usb/cdns3/gadget.c
+++ b/drivers/usb/cdns3/gadget.c
@@ -493,7 +493,7 @@ static int cdns_req_ep0_get_status(struct usb_ss_dev *usb_ss,
 				usb_status |= 1uL << USB_DEVICE_REMOTE_WAKEUP;
 
 			/* self powered */
-			usb_status |= 1uL << USB_DEVICE_SELF_POWERED;
+			usb_status |= usb_ss->gadget.is_selfpowered;
 		}
 		break;
 
@@ -1910,9 +1910,14 @@ static int usb_ss_gadget_set_selfpowered(struct usb_gadget *gadget,
 		int is_selfpowered)
 {
 	struct usb_ss_dev *usb_ss = gadget_to_usb_ss(gadget);
+	unsigned long flags;
 
 	dev_dbg(&usb_ss->dev, "usb_ss_gadget_set_selfpowered: %d\n",
 		is_selfpowered);
+
+	spin_lock_irqsave(&usb_ss->lock, flags);
+	gadget->is_selfpowered = !!is_selfpowered;
+	spin_unlock_irqrestore(&usb_ss->lock, flags);
 	return 0;
 }
 
-- 
1.7.9.5

